### PR TITLE
Add support for etcd over tls transport. 

### PR DESCRIPTION
**Pull request by georgekuruvillak**
Change etcdstatefulset and apiserver to include changes to have tls transport.

### CODE REVIEW

File: charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml]: Can we change that to `IfNotPresent`, please?

File: charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml]: Can we reverse the order (first `name` and then `mountPath`) to improve readability?

**Review by rfranzke — COMMENTED** [review]
Thanks a lot! First quick review done, will do another once you tell me it's ready.

File: pkg/operation/botanist/secrets.go
Comment (mvladev) [inline pkg/operation/botanist/secrets.go]: can we generate a CA for `etcd`? Sharing the `kubernetes` CA is a bad practice.

File: pkg/operation/botanist/secrets.go
Comment (mvladev) [inline pkg/operation/botanist/secrets.go:614]: It's better to use a headless kubernetes service and the API server to talk to `etcd-{role}-0.etcd-{role}`

**Review by mvladev — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/secrets.go
Comment (rfranzke) [reply pkg/operation/botanist/secrets.go]: Maybe we can postpone that until the way how we generate certificates/secrets has been refactored? It's already quite messy and the CA part is not well encapsulated into functions.

**Comment by rfranzke** [comment]
@georgekuruvillak whats the status here?

File: charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml:22]: Just to understand that: The `etcd-<role>-client` service is not needed (because the API server talks to etcd now via the headless service), right? And Prometheus knows how to scrape the etcd because of the static configuration?

File: charts/seed-controlplane/charts/etcd/values.yaml
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/values.yaml]: To be clear that it is the name of a secret, let's add the `Name` suffix to the key.

File: charts/seed-controlplane/charts/etcd/values.yaml
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/values.yaml]: To be clear that it is the name of a secret, let's add the `Name` suffix to the key.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Looks good overall, added a few things inline, PTAL.

File: charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml
Comment (georgekuruvillak) [reply charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml:22]: Yes. We use headless service at port 2379. Service @2380 was for peer communication which we do not need. Prometheus knows how to scrap from etcd instances because we are providing it with the headless endpoints in `scrap_config`.

**Comment by rfranzke** [comment]
OK, looks good now.
Still on hold until a new release of etcd-backup-restore component has been built so that I can test it.

File: pkg/operation/botanist/secrets.go
Comment (georgekuruvillak) [reply pkg/operation/botanist/secrets.go:614]: Why do you suggest that we need the etcd client as a headless service? For peer communication endpoints, it makes sense because the communication has to be directed to a particular instance instead of any instances. For client endpoints, we have the liberty of directing the request to any of the instances (though we have just one).

File: charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml
Comment (georgekuruvillak) [reply charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml:22]: I have switched back to a ClusterIP service from the headless client implementation. Prometheus scraps using service DNS name.

**Comment by rfranzke** [comment]
Something went wrong with your rebase, the PR now covers 31 commits and has a lot of conflicts. Please rebase again so that it contains only your changes.

**Comment by rfranzke** [comment]
Thanks for rebasing, so, just tested this out. I see some error logs in the etcd container:
```
2018-05-23 08:01:49.853328 I | embed: rejected connection from "10.243.147.208:37716" (error "EOF", ServerName "")
2018-05-23 08:01:51.952031 I | embed: rejected connection from "10.243.147.198:49620" (error "tls: first record does not look like a TLS handshake", ServerName "")
2018-05-23 08:01:59.959240 I | embed: rejected connection from "10.243.147.208:37764" (error "EOF", ServerName "")
2018-05-23 08:02:01.451217 I | embed: rejected connection from "10.243.147.198:58666" (error "remote error: tls: bad certificate", ServerName "etcd-main-client.shoot--dev--testcluster.svc")
```

**Comment by rfranzke** [comment]
After a call with George, here is a list of things to be done before we can merge:
- [x] Rebase on current `master`.
- [x] Add checksums to the etcd statefulset pods for the new secrets mounted (ca, server cert, client cert)
- [x] Keep the `etcd-{role}-client` service 
- [x] Delete the `etcd-{role}` service in `DeployETCD` function
- [x] Update etcd version to 3.3.5 (already tested and verified)
- [x] Make the etcd-server cert having both server and client usage (see 
https://coreos.com/etcd/docs/latest/op-guide/security.html#frequently-asked-questions).
- [x] Adapt the kube-apiserver flags to communicate with the `-client` service.
- [x] Change the etcd server certificate SANS to have the `-client` suffix.
- [x] ~We see `2018-05-23 09:26:45.038635 I | embed: rejected connection from "10.243.154.168:36408" (error "EOF", ServerName "")` in the etcd logs whenever the kube-apiservers `/healthz` endpoint is queried. The kube-apiserver itself checks whether it can connect to etcd (https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/rest/storage_core.go#L257-L286). Although etcd reports that the connection was rejected, the kube-apiserver answers with `200 OK` to the `/healthz` request. Upstream etcd issue: https://github.com/coreos/etcd/issues/9285, nothing we can do for now~

**Comment by mvladev** [comment]
> Delete the etcd-{role} service in DeployETCD function

Why remove the headless service? In this case you can't use NS query for `etcd-main-0.etcd-main`

> Make the etcd-server cert having both server and client usage (see
https://coreos.com/etcd/docs/latest/op-guide/security.html#frequently-asked-questions).

Why? Normally you have server certificate and client certificate.

**Comment by georgekuruvillak** [comment]
But why would we need to have a directed query to the instance? It should be ok that the service takes care of the resolution to any instance for a client request. When we go multi-instance we should think of a headless service for peer-to-peer communication.

Etcd requires us to have both capabilities in the same Certificate for client authentication. Surprising but without it, it results in error.

**Comment by mvladev** [comment]
> Etcd requires us to have both capabilities in the same Certificate for client authentication. Surprising but without it, it results in error.

Yes, because there are 3 certificates used for `etcd`:

- **server** with `signing`, `key encipherment` and `server auth` usages
- **client** with `signing`, `key encipherment` and `client auth` usages
- **peering** with `signing`, `key encipherment`, `server auth` and `client auth` usages

If you don't provide the **peering** certificate, then use the **server** certificate.

> But why would we need to have a directed query to the instance? It should be ok that the service takes care of the resolution to any instance for a client request. When we go multi-instance we should think of a headless service for peer-to-peer communication.

One prime example for this is Prometheus - it needs to scrape _ALL_ instances at the same time.

I would suggest to use DNS for etcd [bootstrap](https://coreos.com/etcd/docs/latest/op-guide/clustering.html#dns-discovery 
). KubeDNS and CoreDNS already support [SVC record](https://github.com/kubernetes/dns/blob/master/docs/specification.md#232---srv-records):

For example:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app: nginx
```

returns

```bash
dig +noall +answer SRV _web._tcp.nginx.default.svc.cluster.local
_web._tcp.nginx.default.svc.cluster.local. 30 IN SRV 10 33 80 web-0.nginx.default.svc.cluster.local.
_web._tcp.nginx.default.svc.cluster.local. 30 IN SRV 10 33 80 web-1.nginx.default.svc.cluster.local.
_web._tcp.nginx.default.svc.cluster.local. 30 IN SRV 10 33 80 web-2.nginx.default.svc.cluster.local.
```

**Review by rfranzke — APPROVED** [review]

**Pull request merged**
